pkg/playback/http_replayer.go

```diff
@@ -0,0 +1,198 @@
+	sequentialComparator := func(req1 interface{}, req2 interface{}) (accept bool, shortCircuitNow bool) {
+		return true, true
+	}
+
+	replayer, err := newInteractionReplayer(reader, httpRequestfromBytes, httpResponsefromBytes, sequentialComparator)
+	if err != nil {
+		return err
+	}
+
+	httpReplayer.replayer = replayer
+
+	return nil
+}
+
+func (httpReplayer *replayServer) handler(w http.ResponseWriter, r *http.Request) {
+	httpReplayer.replayLock.Wait()       // wait to make sure no webhooks are in the middle of being fired
+	httpReplayer.replayLock.Add(1)       // acquire the lock so we can handle this request
+	defer httpReplayer.replayLock.Done() // release the lock when handler func is done
+
+	fmt.Printf("\n--> %v to %v", r.Method, r.RequestURI)
+
+	wrappedRequest, err := newHTTPRequest(r)
+	if err != nil {
+		writeErrorToHTTPResponse(w, err, 500)
+		return
+	}
+
+	// --- Read matching response from cassette
+	var wrappedResponse *httpResponse
+	wrappedResponse, err = httpReplayer.getNextRecordedCassetteResponse(&wrappedRequest)
+	if err != nil {
+		writeErrorToHTTPResponse(w, err, 500)
+		return
+	}
+
+	fmt.Printf("\n<-- %v from %v\n", wrappedResponse.StatusCode, "CASSETTE")
+
+	// --- Write response back to client
+	// The header *must* be written first, since writing the body with implicitly and irreversibly set
+	// the status code to 200 if not already set.
+	copyHTTPHeader(w.Header(), wrappedResponse.Headers) // header map must be written before calling w.WriteHeader
+	w.WriteHeader(wrappedResponse.StatusCode)
+	io.Copy(w, bytes.NewBuffer(wrappedResponse.Body))
+
+	// --- Handle webhooks
+	// Check the cassette to see if there are pending webhooks we should fire
+
+	// Note on calling resp.Body().Close() and avoiding a resource leak: all response bodies will be closed in the goroutine below
+	webhookRequests, webhookResponses, err := httpReplayer.readAnyPendingWebhookRecordingsFromCassette() // nolint
+
+	if err != nil {
+		fmt.Printf("Error when checking cassette for webhooks to replay: %v\n", err)
+	}
+	fmt.Printf("Replaying %d webhooks.\n", len(webhookRequests))
+
+	// Send the webhooks
+
+	// Struct used to parse the event.Type from recorded webhook JSON bodies
+	type stripeEvent struct {
+		Type string `json:"type"`
+	}
+
+	go func() {
+		// Note: if there are any errors in processing recorded webhooks here,
+		// we log the error and keep going.
+		// TODO: is the above the appropriate behavior?
+		httpReplayer.replayLock.Wait()       // only send webhooks after the previous request/response is handled
+		httpReplayer.replayLock.Add(1)       // acquire lock so we can send webhooks
+		defer httpReplayer.replayLock.Done() // release lock when done sending webhooks
+
+		for i, webhookReq := range webhookRequests {
+			resp, err := forwardRequest(webhookReq, httpReplayer.webhookURL)
+			if err != nil {
+				fmt.Printf("ERROR when forwarding webhook requests: %v", err)
```

Comment:
Could we perhaps add a unique identifier in the following error messages that call out which request failed to forward? Such as the Stripe API URL called and the method.

Reply:
Ah good idea - done in 3eacd92780930d55afa79d0c12f2ee5e7e0aec41.